### PR TITLE
Add deb/rpm packages

### PR DESCRIPTION
**Pull request by nickatsegment**
- Build deb/rpm packages in `dist` job, using nfpm
- Split `Makefile` in to `Makefile` (simple; mostly aimed at users who want to build it themselves), `Makefile.release` (complex; aimed at CI/devs) and `Makefile.tools` (one-off installation tasks)
- Add `test` job/recipe
- Publish prerelease tags (e.g. `v2.1.0-anythingwithadash`) as github prereleases (mostly for testing this process)
- Publish binaries, packages, and sha256sum to a Github release
- Publish packages to PackageCloud

Ended up refactoring a lot of the Makefile and fixing circle in general. I had to use a fair few GNU Make tricks to get it to work in the end; maybe that's crazy

# Testing deb
```
$ docker run -it -v `pwd`:/tmp debian sh
# cd /tmp
# dpkg -i chamber_v2.1.0-prerelease-test-dev_amd64.deb
dpkg: warning: parsing file '/var/lib/dpkg/tmp.ci/control' near line 12 package 'chamber':
 missing maintainer
Selecting previously unselected package chamber.
(Reading database ... 6498 files and directories currently installed.)
Preparing to unpack chamber_v2.1.0-prerelease-test-dev_amd64.deb ...
Unpacking chamber (2.1.0-prerelease-test-dev) ...
Setting up chamber (2.1.0-prerelease-test-dev) ...
# chamber version
chamber v2.1.0-prerelease-test-dev
```

# Testing rpm
```
$ docker run -it -v `pwd`:/tmp fedora sh
sh-4.4# cd /tmp
sh-4.4# rpm -i chamber_v2.1.0-prerelease-test-dev_amd64.rpm
sh-4.4# chamber version
chamber v2.1.0-prerelease-test-dev
```

# Testing tags
* `git tag v2.1.0-prerelease-test ; git push origin v2.1.0-prerelease-test`
* Triggered this workflow https://circleci.com/workflow-run/17b45616-28d0-49eb-802d-79a7d2936996
* And this release https://github.com/segmentio/chamber/releases/tag/v2.1.0-prerelease-test

# Testing Packagecloud releases
* `git tag -f v2.1.0-prerelease-test3 && git push -f origin v2.1.0-prerelease-test3`
* Triggered this workflow: https://circleci.com/workflow-run/4c05c0de-d48f-45af-9f89-9c96bace1db1
* And this package: https://packagecloud.io/segment/chamber-prerelease/packages/debian/stretch/chamber_2.1.0-prerelease-test3_amd64.deb

# Installing via apt on debian

```
$ docker run -it debian sh -c 'apt update -q && apt install -yq curl && curl -s https://packagecloud.io/install/repositories/segment/chamber-prerelease/script.deb.sh | bash >/dev/null 2>&1 && apt install -yq chamber && chamber version'
...
Get:1 https://packagecloud.io/segment/chamber-prerelease/debian stretch/main amd64 chamber amd64 2.1.0-prerelease-test3 [4896 kB]
Fetched 4896 kB in 1s (4246 kB/s)
debconf: delaying package configuration, since apt-utils is not installed
Selecting previously unselected package chamber.
(Reading database ... 7291 files and directories currently installed.)
Preparing to unpack .../chamber_2.1.0-prerelease-test3_amd64.deb ...
Unpacking chamber (2.1.0-prerelease-test3) ...
Setting up chamber (2.1.0-prerelease-test3) ...
chamber v2.1.0-prerelease-test3
```

# Installing via yum on fedora
```
$ docker run -it fedora sh -c 'curl -s https://packagecloud.io/install/repositories/segment/chamber-prerelease/script.rpm.sh | bash && yum install -y chamber && chamber version'
Detected operating system as fedora/28.
Checking for curl...
Detected curl...
Downloading repository file: https://packagecloud.io/install/repositories/segment/chamber-prerelease/config_file.repo?os=fedora&dist=28&source=script
done.
Installing pygpgme to verify GPG signatures...
Fedora 28 - x86_64 - Updates                                                                                                                        13 MB/s |  24 MB     00:01
Fedora 28 - x86_64                                                                                                                                  16 MB/s |  60 MB     00:03
Importing GPG key 0x06C8B876:
 Userid     : "https://packagecloud.io/segment/chamber-prerelease (https://packagecloud.io/docs#gpg_signing) <support@packagecloud.io>"
 Fingerprint: 2965 CB52 8B88 897F EEE5 6625 90F4 AD12 06C8 B876
 From       : https://packagecloud.io/segment/chamber-prerelease/gpgkey
segment_chamber-prerelease-source                                                                                                                  781  B/s | 296  B     00:00
...

Total download size: 4.6 M
Installed size: 17 M
Downloading Packages:
chamber-2.1.0_prerelease_test4-1.x86_64.rpm                                                                                                        8.2 MB/s | 4.6 MB     00:00
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                              8.1 MB/s | 4.6 MB     00:00
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                           1/1
  Running scriptlet: chamber-2.1.0_prerelease_test4-1.x86_64                                                                                                                   1/1
  Installing       : chamber-2.1.0_prerelease_test4-1.x86_64                                                                                                                   1/1
  Running scriptlet: chamber-2.1.0_prerelease_test4-1.x86_64                                                                                                                   1/1
  Verifying        : chamber-2.1.0_prerelease_test4-1.x86_64                                                                                                                   1/1

Installed:
  chamber.x86_64 2.1.0_prerelease_test4-1

Complete!
chamber v2.1.0-prerelease-test4
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,56 +1,33 @@
...
-	github-release upload \
-	--security-token $$GH_LOGIN \
-	--user segmentio \
-	--repo chamber \
-	--tag $(VERSION) \
-	--name chamber-$(VERSION).sha256sums \
-	--file dist/chamber-$(VERSION).sha256sums
+	go test -v ./...
+
+all: dist/chamber-$(VERSION)-darwin-amd64 dist/chamber-$(VERSION)-linux-amd64
```
Comment (systemizer) [inline Makefile]: assuming we can get rid of this since these are now referenced in other makefiles

File: Makefile.release [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+publish-github: publish-github-darwin publish-github-linux publish-github-deb publish-github-rpm publish-github-sha256sums
+
+publish-packagecloud: publish-packagecloud-deb publish-packagecloud-rpm
+
+github-release:
+	github-release release \
+	--security-token $$GH_LOGIN \
+	--user segmentio \
+	--repo chamber \
+	$(GITHUB_RELEASE_FLAGS) \
```
Comment (systemizer) [inline Makefile.release]: this want to make sure this still works if it's not pre-release (looks like this variable is only set if there isn't a dash in the tag

File: Makefile.release [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+PACKAGECLOUD_DEB_DISTROS := \
+	debian/stretch
+
+PACKAGECLOUD_RPM_DISTROS := \
+	fedora/27 \
+	fedora/28
+
+publish: publish-github publish-packagecloud
+
+publish-github: publish-github-darwin publish-github-linux publish-github-deb publish-github-rpm publish-github-sha256sums
```
Comment (systemizer) [inline Makefile.release]: is it okay if each of these run the github-release target individually? (based on what i read that's what i think what's happening)

Comment (nickatsegment) [reply Makefile.release]: Yep, the magic of Make means it only gets run once

**Comment by nickatsegment** [comment]
After merging, I tested a real-ish prerelease tag and it works :) https://circleci.com/gh/segmentio/chamber/257

**Pull request merged**